intermediate_source/pinmem_nonblock.py

```diff
@@ -0,0 +1,625 @@
+#   - :ref:`non_blocking=True <pinmem_nb>`
+#   - :ref:`Synergies <synergies>`
+#   - :ref:`Other copy directions (GPU -> CPU) <pinmem_otherdir>`
+#
+# - :ref:`Practical recommendations <pinmem_recom>`
+# - :ref:`Additional considerations <pinmem_considerations>`
+# - :ref:`Conclusion <pinmem_conclusion>`
+# - :ref:`Additional resources <pinmem_resources>`
+#
+#
+# Background
+# ----------
+#
+#   .. _pinmem_background:
+#
+# Memory management basics
+# ~~~~~~~~~~~~~~~~~~~~~~~~
+#
+#   .. _pinmem_mem:
+#
+# When one creates a CPU tensor in PyTorch, the content of this tensor needs to be placed
+# in memory. The memory we talk about here is a rather complex concept worth looking at carefully.
+# We distinguish two types of memories that are handled by the Memory Management Unit: the main memory (for simplicity)
+# and the disk (which may or may not be the hard drive). Together, the available space in disk and RAM (physical memory)
+# make up the virtual memory, which is an abstraction of the total resources available.
+# In short, the virtual memory makes it so that the available space is larger than what can be found on RAM in isolation
+# and creates the illusion that the main memory is larger than it actually is.
+#
+# In normal circumstances, a regular CPU tensor is _paged_, which means that it is divided in blocks called _pages_ that
+# can live anywhere in the virtual memory (both in RAM or on disk). As mentioned earlier, this has the advantage that
+# the memory seems larger than what the main memory actually is.
+#
+# Typically, when a program accesses a page that is not in RAM, a "page fault" occurs and the operating system (OS) then brings
+# back this page into RAM (_swap in_ or _page in_).
+# In turn, the OS may have to _swap out_ (or _page out_) another page to make room for the new page.
+#
+# In contrast to pageable memory, a _pinned_ (or _page-locked_ or _non-pageable_) memory is a type of memory that cannot
+# be swapped out to disk.
+# It allows for faster and more predictable access times, but has the downside that it is more limited than the
+# pageable memory (aka the main memory).
+#
+# .. figure:: /_static/img/pinmem.png
+#    :alt:
+#
+# CUDA and (non-)pageable memory
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+#   .. _pinmem_cuda_pageable_mem:
+#
+# To understand how CUDA copies a tensor from CPU to CUDA, let's consider the two scenarios above:
+# - If the memory is page-locked, the device can access the memory directly in the main memory. The memory addresses are well
+#   defined and functions that need to read these data can be significantly accelerated.
+# - If the memory is pageable, all the pages will have to be brought to the main memory before being sent to the GPU.
+#   This operation may take time and is less predictable than when executed on page-locked tensors.
+#
+# More precisely, when CUDA sends pageable data from CPU to GPU, it must first create a page-locked copy of that data
```

Comment:
and "page-locked" is the same thing as "pinned"?

Reply:
That's my understanding!